src/overlaybd/tar/erofs/test/erofs_stress_base.cpp

```diff
@@ -0,0 +1,410 @@
+	if (system(cmd.c_str()))
+		LOG_ERROR_RETURN(-1, false, "fail to prepare tar file, cmd: `", cmd);
+	prefix = origin_prefix;
+	return true;
+}
+
+LSMT::IFileRW *StressBase::mkfs()
+{
+	LSMT::IFileRW *lowers = nullptr;
+	for (int i = 0; i < num_layers; i ++) {
+		LOG_INFO("processing layer `", i);
+		std::string src_path = prefix + "/layer" + std::to_string(i) + ".tar";
+		std::string idx_path = prefix + "/layer" + std::to_string(i) + ".idx";
+		std::string meta_path = prefix + "/layer" + std::to_string(i) + ".meta";
+
+		/* prepare for idx and meta files */
+		auto src_file = host_fs->open(src_path.c_str(), O_RDONLY, 0666);
+		auto idx_file = host_fs->open(idx_path.c_str(), O_RDWR | O_CREAT | O_TRUNC, S_IRWXU);
+		auto meta_file = host_fs->open(meta_path.c_str(), O_RDWR | O_CREAT | O_TRUNC, S_IRWXU);
+		if (!src_file || !idx_file || !meta_file)
+			LOG_ERROR_RETURN(-1, nullptr, "fail to prepare tar, idx or meta file for layer `", std::to_string(i));
+
+		LSMT::WarpFileArgs args(idx_file, meta_file, src_file);
+		args.virtual_size = IMAGE_SIZE;
+		LSMT::IFileRW *current_layer = create_warpfile(args, false);
+		if (!current_layer)
+			LOG_ERROR_RETURN(-1, nullptr, "fail to prepare wrapfile for layer `", std::to_string(i));
+
+		LSMT::IFileRW *img_file = nullptr;
+		if (i > 0)
+			img_file = LSMT::stack_files(current_layer, lowers, false, false);
+		else
+			img_file = current_layer;
+
+		/* create erofs fs image */
+		auto tar = new LibErofs(img_file, 4096, false);
+		if (tar->extract_tar(src_file, true, i == 0)) {
+			delete img_file;
+			LOG_ERROR_RETURN(-1, nullptr, "fail to extract tar");
+		}
+		delete lowers;
+		delete tar;
+		lowers = img_file;
+	}
+
+	return lowers;
+}
+
+bool StressBase::verify(photon::fs::IFileSystem *erofs_fs) {
+	std::vector<std::string> items;
+	std::string cur;
+	struct stat st;
+	bool first = true;
+
+	items.emplace_back(std::string("/"));
+	do {
+		StressNode *node;
+		cur = items.front();
+		items.erase(items.begin());
+
+		if (erofs_fs->stat(cur.c_str(), &st))
+			LOG_ERRNO_RETURN(-1, false, "fail to stat file `", cur);
+		if (S_ISDIR(st.st_mode)) {
+			node = new StressNode(cur, NODE_DIR);
+			auto dir = erofs_fs->opendir(cur.c_str());
+			/* the dir may be empty, so check it first */
+			if (dir->get() != nullptr) {
+				do {
+					dirent *dent = dir->get();
+
+					if (first)
+						items.emplace_back(cur + std::string(dent->d_name));
+					else
+						items.emplace_back(cur + "/" + std::string(dent->d_name));
+				} while (dir->next());
+			}
+			dir->closedir();
+			delete dir;
+		} else if (S_ISREG(st.st_mode)) {
+			photon::fs::IFile *file;
+			bool ret;
+
+			file = erofs_fs->open(cur.c_str(), O_RDONLY);
+			node = new StressNode(cur, NODE_REGULAR);
+			if (!file || ! node)
+				LOG_ERROR_RETURN(0, false, "fail to open file or node `", cur);
+			ret = verify_gen_mod(node, file) &&
```

Comment:
it should check if the whole stat() equals.

Reply:
Done.